File: Algorithms/146_LruCache.py
class Node(object):
    def __init__(self, k=None, v=None):
        self.key = k
        self.value = v
        self.pre = None
        self.next = None

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity, self.size = capacity, 0
        self.m = {}
        self.head, self.tail = Node(), Node()
        self.head.next = self.tail
        self.tail.pre = self.head

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.m:
            p = self.m[key]
            p.next.pre = p.pre
            p.pre.next = p.next
            p.next = self.head.next
            self.head.next.pre = p
            p.pre = self.head
            self.head.next = p
            return p.value
        else:
            return -1


    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if key in self.m:
            p = self.m[key]
            p.next.pre = p.pre
            p.pre.next = p.next
            p.next = self.head.next
            self.head.next.pre = p
            p.pre = self.head
            self.head.next = p
            p.value = value
        else:
            p = Node(key, value)
            p.next = self.head.next;
            self.head.next.pre = p
            p.pre = self.head
            self.head.next = p
            self.m[key] = p
            self.size += 1
            if self.size > self.capacity:
                p = self.tail.pre
                p.pre.next = self.tail
                self.tail.pre = p.pre
                del self.m[p.key]
                self.size -= 1

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: Algorithms/java/146.LruCache.java
public class LRUCache {
    class Node {
        int key, value;
        Node pre, next;
        Node(int k, int v) {
            key = k;
            value = v;
            pre = null;
            next = null;
        }
    }
    public Node head, tail;
    public int capacity;
    public int size;
    public HashMap<Integer, Node> m;
    public LRUCache(int capacity) {
        this.capacity = capacity;
        size = 0; 
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head.next = tail;
        tail.pre = head;
        m = new HashMap<>();
    }
    
    public int get(int key) {
        if (m.containsKey(key)) {
            Node p = m.get(key);
            p.next.pre = p.pre;
            p.pre.next = p.next;
            p.next = head.next;
            head.next.pre = p;
            p.pre = head;
            head.next = p;
            return p.value;
        } else {
            return -1;
        }    
    }
    
    public void put(int key, int value) {
       if (m.containsKey(key)) {
            Node p = m.get(key);
            p.next.pre = p.pre;
            p.pre.next = p.next;
            p.next = head.next;
            head.next.pre = p;
            p.pre = head;
            head.next = p;
            p.value = value;
       } else {
           Node p = new Node(key, value);
           p.next = head.next;
           head.next.pre = p;
           p.pre = head;
           head.next = p;
           m.put(key, p);
           size ++;
           if (size > capacity) {
               p = tail.pre;
               p.pre.next = tail;
               tail.pre = p.pre;
               m.remove(p.key);
               size --;
           }
       } 
    }
}
